### PR TITLE
[Merged by Bors] - feat: Mul by invertible matrices is injective even for rectangular matrices

### PR DESCRIPTION
**Pull request by MohanadAhmed**
Multiplication by invertible matrices from the left or right is injective. Note that [`mul_left_injective₀`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupWithZero/Defs.html#mul_left_injective%E2%82%80) and friends don't apply because they would require similar (square) matrices.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean [hunk 1]
```diff
@@ -356,6 +356,16 @@ theorem mul_inv_eq_iff_eq_mul_of_invertible (A B C : Matrix n n α) [Invertible
    fun h => by rw [h, mul_inv_cancel_right_of_invertible]⟩
 #align matrix.mul_inv_eq_iff_eq_mul_of_invertible Matrix.mul_inv_eq_iff_eq_mul_of_invertible
 
+lemma mul_left_injective_of_invertible
+    {m: Type} [Fintype m] [DecidableEq m] (P : Matrix m m α) [Invertible P] :
+    Function.Injective (fun (x : Matrix m n α) => P ⬝ x) :=
+  fun _ _ h => by simpa only [inv_mul_cancel_left_of_invertible] using (congr_arg (P⁻¹ ⬝ ·) h)
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean:362]: This should be called
```suggestion
lemma mul_right_injective_of_invertible
    {m: Type} [Fintype m] [DecidableEq m] (P : Matrix m m α) [Invertible P] :
    Function.Injective (fun (x : Matrix m n α) => P ⬝ x) :=
  fun _ _ h => by simpa only [inv_mul_cancel_left_of_invertible] using (congr_arg (P⁻¹ ⬝ ·) h)
```
to match the order of `units.mul_right_inj`.

Can you also add `mul_right_inj_of_invertible`, which can be proved as `(mul_right_injective_of_invertible P).eq_iff`?

File: Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean [hunk 2]
```diff
@@ -356,6 +356,16 @@ theorem mul_inv_eq_iff_eq_mul_of_invertible (A B C : Matrix n n α) [Invertible
...
+
+lemma mul_right_injective_of_invertible
+    {m: Type} [Fintype m] [DecidableEq m] (P : Matrix m m α) [Invertible P] :
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean:365]: Is there an ambient `m` you can use? This one is not universe polymorphic!

File: Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean [hunk 3]
```diff
@@ -356,6 +356,20 @@ theorem mul_inv_eq_iff_eq_mul_of_invertible (A B C : Matrix n n α) [Invertible
    fun h => by rw [h, mul_inv_cancel_right_of_invertible]⟩
 #align matrix.mul_inv_eq_iff_eq_mul_of_invertible Matrix.mul_inv_eq_iff_eq_mul_of_invertible
 
+lemma mul_right_injective_of_invertible [Invertible A] :
+    Function.Injective (fun (x : Matrix n m α) => A ⬝ x) :=
+  fun _ _ h => by simpa only [inv_mul_cancel_left_of_invertible] using (congr_arg (A⁻¹ ⬝ ·) h)
+
+lemma mul_left_injective_of_invertible [Invertible A] :
+    Function.Injective (fun (x : Matrix m n α) => x ⬝ A) :=
+  fun a x hax => by simpa only [mul_inv_cancel_right_of_invertible] using (congr_arg (· ⬝ A⁻¹) hax)
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean:365]: nit:
```suggestion
lemma mul_right_injective_of_invertible [Invertible A] :
    Function.Injective (fun x : Matrix n m α => A ⬝ x) :=
  fun _ _ h => by simpa only [inv_mul_cancel_left_of_invertible] using congr_arg (A⁻¹ ⬝ ·) h

lemma mul_left_injective_of_invertible [Invertible A] :
    Function.Injective (fun x : Matrix m n α => x ⬝ A) :=
  fun a x hax => by simpa only [mul_inv_cancel_right_of_invertible] using congr_arg (· ⬝ A⁻¹) hax
```

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

**Comment by MohanadAhmed** [comment]
> ✌️ MohanadAhmed can now approve this pull request. To approve and merge a pull request, simply reply with `bors r+`. More detailed instructions are available [here](https://bors.tech/documentation/getting-started/#reviewing-pull-requests).

bors r+

**Comment by MohanadAhmed** [comment]
> > ✌️ MohanadAhmed can now approve this pull request. To approve and merge a pull request, simply reply with `bors r+`. More detailed instructions are available [here](https://bors.tech/documentation/getting-started/#reviewing-pull-requests).
> 
> bors r+

bors r+

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**